File: Algorithms/C++/Backtracking/39_Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> combinations;

        if (candidates.empty()) {
            return combinations;
        }

        sort(candidates.begin(), candidates.end());
        vector<int> cur;

        helper(combinations, cur, 0, candidates, target);

        return combinations;
    }
private:
    void helper(vector<vector<int>>& combinations, vector<int>& cur, int pos, vector<int>& candidates, int target) {
        if (target == 0) {
            combinations.push_back(cur);
            return;
        }

        for (int i = pos; i < candidates.size(); i++) {
            if (candidates[i] > target) {
                break;
            }
            cur.push_back(candidates[i]);
            helper(combinations, cur, i, candidates, target - candidates[i]);
            cur.pop_back();
        }
    }
};

File: Algorithms/Swift/Backtracking/39_Combination_Sum.swift
class Solution {
    func combinationSum(candidates: [Int], _ target: Int) -> [[Int]] {
        var combinations = [[Int]]()

        if candidates.isEmpty {
            return combinations
        }

        var sortedCandidates = candidates
        sortedCandidates.sortInPlace()

        var currentCombination = [Int]()
        helper(&combinations, &currentCombination, 0, sortedCandidates, target)

        return combinations
    }

    private func helper(inout combinations: [[Int]], inout _ currentCombination: [Int], _ position: Int, _ candidates: [Int], _ target: Int) {
        if target == 0 {
            combinations.append(currentCombination)
            return
        }

        var previous: Int! = nil
        for var i = position; i < candidates.count; i++ {
            if candidates[i] > target {
                return
            }

            if previous != nil && previous == candidates[i] {
                continue
            }

            currentCombination.append(candidates[i])
            helper(&combinations, &currentCombination, i, candidates, target - candidates[i])
            currentCombination.removeLast()

            previous = candidates[i]
        }
    }
}
